### PR TITLE
add rest api to send the arguments template for request

### PR DESCRIPTION
**Pull request by ticoann**

### CODE REVIEW

**Comment by amaltaro** [comment]
You have set 'ConfigCacheUrl ' optional for some request types and mandatory for others.
I would rather make it optional and assign cmsweb production url as default value.
More details:
 - MonteCarlo.py: set optional to True
 - MonteCarloFromGEN.py: assign a default value
 - ReDigi.py: add this argument to the specArgs
 - ReReco.py: set optional to True
 - TaskChain.py: assign a default value and set null to False. I would also remove ConfigCacheUrl from the getTaskArguments(), I guess we can assume the steps configuration will always be available in the same couch backend (?)

**Comment by ticoann** [comment]
Hi Alan, 
I think we are using optional confusing way. It seems optional means whether user can specify or not instead of whether the parameter itself is optional in the spec.
I added optional False for ConfigCacheUrl when ConfigCacheID is non optional.

I would like to change the code do following.
1. If non optional (mandatory), and default value exists. 
    User can overwrite the value, but if user doesn't specify the value, it takes default value.
2. If non optional and default value doesn't exist (None). 
    User must specify the value. Otherwise validation failes.
3. For the optional parameter, default value shouldn't exist.

**Comment by amaltaro** [comment]
> It seems optional means whether user can specify or not instead of whether the parameter itself is optional in the spec.

This was my understanding as well (just replacing "can" by "must" :-D).

> I added optional False for ConfigCacheUrl when ConfigCacheID is non optional.

ConfigCacheID is always non optional for the ordinary central production specs (MC, TaskChain, ReDigi and ReReco), no? The only case it *could* be optional - I guess - would be for Tier0 wfs, no?

About your 3 points, I agree with 1) and 2) but not 3). There are a few optional parameters where it's useful to have a default value (e.g., KeepOutput, DeterministicPileup, etc).
What do you think? :-D

**Comment by hufnagel** [comment]
>     I added optional False for ConfigCacheUrl when ConfigCacheID is non
>     optional.
>
> ConfigCacheID is always non optional for the ordinary central production
> specs (MC, TaskChain, ReDigi and ReReco), no? The only case it /could/
> be optional - I guess - would be for Tier0 wfs, no?

It's optional even for ordinary central production. You *can* run
a ReReco using the Config.DP defined configuration if you want.
No one does it, but we support it. This is true for any other
type of workflow as well, in principle if the release provides
the config, it can be used (without ConfigCache).

Tier0 doesn't use any ConfigCache at all, all our configurations
are release defined.

Cheers

	Dirk

**Comment by amaltaro** [comment]
Ok, and my dummy question now. How would we configure a workflow to use this Config.DP?

If a workflow (central production) is injected into the system without a ConfigCacheID, would it (jobs) then use this Config.DP?

**Comment by hufnagel** [comment]
Hm, looking at the code, there is low level support for scenario based processing via Config.DP provided configurations. But the only specs that actually use it are the specs in the T0 repo and the PromptReco spec.

The way it's used is to pass three parameters (scenarioName, scenarioFunc, scenarioArgs) to setupProcessingTask, so this needs to be implemented as on option in the higher level spec.

So looks like it's only supported for Tier0 right now. Not because the low level code doesn't support more, but because we haven't implemented the hooks for more. Still, it means that we don't want to completely rule this out and therefor ConfigCache should remain optional. It's an operational choice that we at the moment don't support anything else, we easily could if we want to (and if there is some future processing where the configs are stable enough to go into a release).

Btw, we use the low level Config.DP code all the time because merge configurations come from Config.DP. So it's not like this is only used in the Tier0.

**Comment by vkuznet** [comment]
Seangchan, you also may fill out Requestor field by capturing cherrypy.request.user headers. I found that it is empty on web UI. Also, I think Input Dataset should be mandatory, so we'll need to add REPLACE- prefix for it.

**Comment by hufnagel** [comment]
InputDataset is definitely not mandatory, at least not in general. You can make it mandatory in derived classes where appropriate, but StdBase should only declare mandatory what is needed *everywhere*.

**Comment by hufnagel** [comment]
Also, we are probably talking past each other here. There are spec parameters which have certain optional/mandatory requirements and then there are Request parameters which map for the most part directly to spec parameters but not 100% one-to-one AFAIK.

**Comment by ticoann** [comment]
I think there is mostly conceptual difference than actual behavior difference between current code and what I suggested.
Current code 
1. has default value for every parameter.
2. if it is none optional default value ignored and user always has to specify the value.
3. if it is optional, it takes default value even though that parameter never used anywhere else.
4. the property of optional values can always be overwritten in inheriting class.

To do the same thing from what I am suggesting.
All values are non optional, 
1. If user need to specify, there is no default value.
2. If user doesn't need to specify, there is default value.

After discussing with Alan. 
I think I will do following.
1. We keep the optional meaning as it is (optional means user doesn't have to specify)
2. We keep the only mandatory parameters (not user mandatory but spec mandatory which could have default value) in StdSpec
3. All inheriting class maintains it own mandatory (spec mandatory) parameters.
So there is no optional parameters for spec, optional means only user optional.

Dirk, Valentin, is that OK?
Dirk could you look over what need to be removed from StdBase class which Tier0 Doesn't use?
Thanks,

**Comment by vkuznet** [comment]
Seangchan,
I'm fine with proposal, but I'd like to clarify things about mandatory
parameters. The API will provide spec mandatory ones, but on web interface
users may need to see which fields they should fill-out. Will you
provide REPLACE- prefix for user mandatory parameters?
Valentin.

**Comment by ticoann** [comment]
> Seangchan, you also may fill out Requestor field by capturing cherrypy.request.user headers. I found that it is empty on web UI. 

Requestor is originally different from user. So someone create the request would be different from some ask to create the request (PPD), The same for RequestorDN, but I don't think we are using that parameter correctly. I will double check with ops.

**Comment by ticoann** [comment]
> on web interface users may need to see which fields they should fill-out. Will you provide REPLACE- prefix for user mandatory parameters?

Yes, for the web interface. If it is mandatory (user), it will have REPLACE- prefix. Otherwise it will still send the parameters with default values.

**Comment by ticoann** [comment]
I think there is mostly conceptual difference than actual behavior difference between current code and what I suggested.
Current code 
1. has default value for every parameter.
2. if it is none optional default value ignored and user always has to specify the value.
3. if it is optional, it takes default value even though that parameter never used anywhere else.
4. the property of optional values can always be overwritten in inheriting class.

To do the same thing from what I am suggesting.
All values are non optional, 
1. If user need to specify, there is no default value.
2. If user doesn't need to specify, there is default value.

**Comment by hufnagel** [comment]
To check what Tier0 is using from StdBase, you could just look at

https://github.com/hufnagel/T0/blob/33fb00111b6a6923ca1e7ea0248990bae2c3228d/src/python/T0/RunConfig/RunConfigAPI.py

and search for specArguments['...'] = ... lines. At the moment, the only redundant parameters are

 # not used, but needed by the validation
specArguments['Group'] = "unknown"
specArguments['Requestor'] = "unknown"
specArguments['RequestorDN'] = "unknown"
specArguments['RequestType'] = "unknown"
specArguments['RequestPriority'] = 0
specArguments['CouchURL'] = "http://fakehost:1234"
specArguments['CouchWorkloadDBName'] = "db_fakename"

which I only set because the validation forces me to. Although you could argue that at least RequestPriority should be correctly set by the Tier0 as well.

**Comment by ticoann** [comment]
Thanks Dirk, I will modify the code accordingly.

**Comment by ticoann** [comment]
Alan, Dirk Could you review the code?

**Comment by hufnagel** [comment]
Design wise ok, but not sure Requestor, RequestorDN and Group can really be optional for me. These parameters *ARE* used in StdBase, they are just unknown for the Tier0. I would take them out and let me override them as before. The Couch settings are fine to be optional for Tier0 only, as I really don't use them.

I'll also override RequestPriority with something sensible.

Btw, the way you have it now, even though PromptReco is a generic spec, this change basically converts it into a spec that can only be used from the Tier0...

**Comment by hufnagel** [comment]
If nothing else I would need to do a full Tier0 replay to check if the new defaults for Requestor etc will work...

File: src/python/WMCore/WMSpec/StdSpecs/StdBase.py [hunk 1]
```diff
@@ -956,7 +948,7 @@ def getWorkloadArguments():
                      "AutoApproveSubscriptionSites" : {"default" : [], "type" : makeList, "assign_optional": False, 
                                         "validate" : lambda x: all([cmsname(y) for y in x])},
                      # should be Low, Normal, High
-                     "SubscriptionPriority" : {"default" : "Low", "type" : str, "assign_optional": False,
+                     "SubscriptionPriority" : {"default" : "Low", "assign_optional": False,
```
Comment (amaltaro) [inline src/python/WMCore/WMSpec/StdSpecs/StdBase.py]: SubscriptionPriority could be optional. At least RelVals don't use this parameter during assignment.

**Comment by amaltaro** [comment]
I didn't look at all bits, but in general it looks ok. Anyways this PR deserves a careful testing :-)
I have 2 other comments, but these would need another PR I guess, they are:
 - it seems CouchURL and ConfigCacheUrl have the same purpose. Perhaps we could remove CouchURL from the code (?)
 - AFAIK DataProcessing is no longer supported, how about we also clean up that one?

**Comment by ticoann** [comment]
> it seems CouchURL and ConfigCacheUrl have the same purpose. Perhaps we could remove CouchURL from the code (?)

Actually there are 2 different things. 
CouchURL is base url for reqmgr db and ConfigCacheUrl is base url for config cache.
Although it is hosted at the same places. When we run the test we usually use our own reqmgr but config cache from cmsweb or testbed.

> AFAIK DataProcessing is no longer supported, how about we also clean up that one?

This is base class which is inherited from others. (ReReco, etc)

File: src/python/WMCore/WMSpec/StdSpecs/StdBase.py (hunk 1 above)
Comment (ticoann) [inline src/python/WMCore/WMSpec/StdSpecs/StdBase.py]: If it is optional, default value "Low" is OK,  I have check how assign_optional handled. Thanks,

**Comment by hufnagel** [comment]
Code doesn't work, some of the changes break things. For instance

"DQMUploadProxy" : {"default" : None, "null" : True, "attr" : "dqmUploadProxy"}

works and correctly fills self.dqmUploadProxy. The new

"DQMUploadProxy" : {"null" : True}

does not. I would highly recommend to always fill at least the attr field for parameters.

**Comment by ticoann** [comment]
Sorry Dirk, it is updated now. 
attr field is automatically generated, but there are some attribute breaking the standard generation rule (change lower case in first letter). I missed DQM case, sorry about that.

**Comment by hufnagel** [comment]
Repack and Express work now, but PromptReco doesn't

2015-01-13 16:08:02,733:ERROR:BaseWorkerThread:Error in worker algorithm (1):
Backtrace:
  <T0Component.Tier0Feeder.Tier0FeederPoller.Tier0FeederPoller instance at 0x1c94dd0> getWorkloadArguments() takes no arguments (1 given)  File "/data/tier0/srv/wmagent/1.9.91/sw.pre.hufnagel/slc6_amd64_gcc481/cms/wmagent/1.0.0.patch3/lib/python2.6/site-packages/WMCore/WorkerThreads/BaseWorkerThread.py", line 172, in __call__
    self.algorithm(parameters)
  File "/data/tier0/srv/wmagent/1.9.91/sw.pre.hufnagel/slc6_amd64_gcc481/cms/t0/1.9.91/lib/python2.6/site-packages/T0Component/Tier0Feeder/Tier0FeederPoller.py", line 195, in algorithm
    self.dqmUploadProxy)
  File "/data/tier0/srv/wmagent/1.9.91/sw.pre.hufnagel/slc6_amd64_gcc481/cms/t0/1.9.91/lib/python2.6/site-packages/T0/RunConfig/RunConfigAPI.py", line 892, in releasePromptReco
    factory = PromptRecoWorkloadFactory()
  File "/data/tier0/srv/wmagent/1.9.91/sw.pre.hufnagel/slc6_amd64_gcc481/cms/wmagent/1.0.0.patch3/lib/python2.6/site-packages/WMCore/WMSpec/StdSpecs/StdBase.py", line 39, in __init__
    argumentDefinition = self.getWorkloadArguments()
  File "/data/tier0/srv/wmagent/1.9.91/sw.pre.hufnagel/slc6_amd64_gcc481/cms/wmagent/1.0.0.patch3/lib/python2.6/site-packages/WMCore/WMSpec/StdSpecs/PromptReco.py", line 286, in getWorkloadArguments
    baseArgs = StdBase.getWorkloadArguments(useReqMgr = False)

**Comment by ticoann** [comment]
Fix for PromptReco and TaskChain is updated now. Please let me know whether other failure occurs.
Sorry for all the trouble.

**Comment by hufnagel** [comment]
PromptReco works now. In fact, together with the T0 pull request (which I have already merged) I see no problems now and I could also remove some of the previous fake parameter overrides I had to do from the Tier0 side.

**Comment by ticoann** [comment]
Thanks Dirk, Alan is running his test now. After his confirmation I will merge the code.

**Comment by amaltaro** [comment]
@ticoann basic tests and all the usual production spec files were exercised. Everything looks fine!
Now I'm going to redeploy WMA (@hufnagel patch) and run a few tests against your VM again, ok?
Thanks

**Comment by ticoann** [comment]
Thank you very much Alan

**Pull request merged**